src/ffi/spatialmath/quaternion.rs

```diff
@@ -0,0 +1,147 @@
+    quat.real = real
+}
+
+/// Set the i component of an existing quaternion stored at the address
+/// of a pointer
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_i(quat_ptr: *mut Quaternion, i: f64) {
+    null_pointer_check!(quat_ptr);
+    let quat = &mut*quat_ptr;
+    quat.i = i
+}
+
+/// Set the j component of an existing quaternion stored at the address
+/// of a pointer
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_j(quat_ptr: *mut Quaternion, j: f64) {
+    null_pointer_check!(quat_ptr);
+    let quat = &mut*quat_ptr;
+    quat.j = j
+}
+
+/// Set the k component of an existing quaternion stored at the address
+/// of a pointer
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_k(quat_ptr: *mut Quaternion, k: f64) {
+    null_pointer_check!(quat_ptr);
+    let quat = &mut*quat_ptr;
+    quat.k = k;
+}
+
+/// Set all of the components of an existing quaternion stored at the address
+/// of a pointer
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_components(
+    quat_ptr: *mut Quaternion, real: f64, i: f64, j: f64, k: f64
+) {
+    null_pointer_check!(quat_ptr);
+    let quat = &mut*quat_ptr;
+    quat.real = real;
+    quat.i = i;
+    quat.j = j;
+    quat.k = k;
+}
+
+/// Set the imaginary components of an existing quaternion stored at
+/// the address of a pointer (quat_ptr) from the components of a 3-vector
+/// (stored at vec_ptr). The convention is x -> i, y -> j, z -> k
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_imag_from_vector(quat_ptr: *mut Quaternion, vec_ptr: *const Vector3) {
+    null_pointer_check!(quat_ptr);
+    null_pointer_check!(vec_ptr);
+    let mut quat = *quat_ptr;
+    let imag = *vec_ptr;
+    quat.set_imag_from_vector(imag);
+}
+
+/// Copies the imaginary components to a 3-vector (using x -> i, y -> j
+/// z -> k) and returns a pointer to the memory address of the resulting
+/// vector
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_get_imaginary_vector(quat_ptr: *const Quaternion) -> *mut Vector3 {
+    null_pointer_check!(quat_ptr);
+    let quat = *quat_ptr;
+    let imag = quat.imag();
+    imag.to_raw_pointer()
+}
+
+/// Converts from euler angles to a quaternion. The euler angles are expected to
+/// be represented according to the Tait-Bryan formalism and applied in the Z-Y'-X"
+/// order (where Z -> yaw, Y -> pitch, X -> roll)
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_from_euler_angles(roll: f64, pitch: f64, yaw: f64) -> *mut Quaternion {
+    let quat = Quaternion::from_euler_angles(roll, pitch, yaw);
+    to_raw_pointer(&quat)
+}
+
+/// Converts a quaternion into euler angles. The euler angles are 
+/// represented according to the Tait-Bryan formalism and applied 
+/// in the Z-Y'-X" order (where Z -> yaw, Y -> pitch, X -> roll). 
+/// The return value is a pointer to a list of [roll, pitch, yaw]
+/// as C doubles
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_to_euler_angles(quat_ptr: *const Quaternion) -> *const c_double {
```

Comment:
is thera a function to free this?

Reply:
to free the quaternion? Or the euler angles?

Reply:
to free the *const c_double, I couldn't find the proper fucntion